### PR TITLE
[NUI] Add webview context and cookiemanager apis

### PR DESCRIPTION
**Pull request by zhouleonlei**
### Description of Change ###
The patch is base on the dali-csharp-binder patches:
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-csharp-binder/+/255352/
https://review.tizen.org/gerrit/#/c/platform/core/uifw/dali-csharp-binder/+/256681

### API Changes ###

 - ACR:

### CODE REVIEW

File: src/Tizen.NUI/src/internal/Interop/Interop.WebContext.cs [hunk 1]
```diff
@@ -33,7 +33,7 @@ internal static partial class WebContext
             [global::System.Runtime.InteropServices.DllImport(NDalicPINVOKE.Lib, EntryPoint = "CSharp_Dali_WebContext_SetCertificateFilePath")]
             public static extern void SetCertificateFilePath(global::System.Runtime.InteropServices.HandleRef jarg1, string jarg2);
 
-            [global::System.Runtime.InteropServices.DllImport(NDalicPINVOKE.Lib, EntryPoint = "CSharp_Dali_WebContext_DisableCache")]
+            [global::System.Runtime.InteropServices.DllImport(NDalicPINVOKE.Lib, EntryPoint = "CSharp_Dali_WebContext_EnableCache")]
             public static extern void DisableCache(global::System.Runtime.InteropServices.HandleRef jarg1, bool jarg2);
```
Comment (elishateng) [inline src/Tizen.NUI/src/internal/Interop/Interop.WebContext.cs:37]: EnableCache vs DisableCache ?

Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/Interop/Interop.WebContext.cs]: > EnableCache vs DisableCache ?

enableCache. Fixed

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 2]
```diff
@@ -164,16 +164,16 @@ public string CertificateFilePath
         /// Disable cache or not.
         /// </summary>
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public bool DisableCache
+        public bool EnableCache
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:167]: EnableCache --> CacheEnabled

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 3]
```diff
@@ -164,16 +164,16 @@ public string CertificateFilePath
...
         {
             get
             {
-                return disableCache;
+                return enableCache;
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:171]: repace 'enableCache'  with Interop.WebContext.IsCacheEnabled(SwigCPtr);

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 4]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
             if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
         }
 
+        /// <summary>
+        /// Get cache enabled.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public bool IsCacheEnabled()
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:349]: removed.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 5]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            bool ret = Interop.WebContext.IsCacheEnabled(SwigCPtr);
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+            return ret;
+        }
+
+        /// <summary>
+        /// Get context certificate file.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public string GetContextCertificateFile()
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:360]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 6]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+            return ret;
+        }
+
+        /// <summary>
+        /// Set context app id.
+        /// <param name="Id">Context app id</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void SetContextAppId(string Id)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:372]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 7]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+            return ret;
+        }
+
+        /// <summary>
+        /// Set context app type.
+        /// <param name="type">Context app type</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void SetContextApplicationType(int type)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:395]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 8]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            Interop.WebContext.SetContextApplicationType(SwigCPtr, type);
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// Set context time offset.
+        /// <param name="offset">Time offset</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void SetContextTimeOffset(float offset)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:406]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 9]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// Set context time zone offset.
+        /// <param name="offset">Time offset</param>
+        /// <param name="time">Daylight saving time</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void SetContextTimeZoneOffset(float offset, float time)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:418]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 10]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// Register url schemes enabled.
+        /// <param name="strArray">The string array of schemes</param>
+        /// <param name="size">The size of schemes</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void RegisterUrlSchemesAsCorsEnabled(string[] strArray, uint size)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:430]: 'uint size; can be removed.
use strArray.length

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 11]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+            Interop.WebContext.RegisterJsPluginMimeTypes(SwigCPtr, strArray, size);
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// Set default zoom factor.
+        /// <param name="value">Default zoom factor</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void SetDefaultZoomFactor(float value)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:453]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 12]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+        {
+            Interop.WebContext.DeleteAllFormCandidateData(SwigCPtr);
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// Get context proxy.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public string GetContextProxy()
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:528]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 13]
```diff
@@ -342,6 +342,230 @@ public void RegisterMimeOverriddenCallback(MimeOverriddenCallback callback)
...
+        {
+            Interop.WebContext.SetContextProxy(SwigCPtr, proxy, rule);
+            if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+        }
+
+        /// <summary>
+        /// get proxy bypass rule.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public string GetProxyBypassRule()
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:551]: It should be a property, not a method.

File: src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs [hunk 14]
```diff
@@ -29,6 +31,13 @@ internal WebCookieManager(global::System.IntPtr cPtr, bool cMemoryOwn) : base(cP
         {
         }
 
+        /// <summary>
+        /// The callback function that is invoked when watch changed.
+        /// </summary>
+        [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public delegate void ChangesWatchCallback();
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs:39]: ChangesWatchCallback --> ChangesWatchedCallback

File: src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs [hunk 15]
```diff
@@ -108,5 +117,17 @@ public void ClearCookies()
             Interop.WebCookieManager.ClearCookies(SwigCPtr);
             if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
         }
+
+        /// <summary>
+        /// Registers callback for changing watch.
+        /// <param name="callback">callback for changing watch</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void RegisterChangesWatchCallback(ChangesWatchCallback callback)
```
Comment (huayongxu) [inline src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs:126]: RegisterChangesWatchCallback --> RegisterChangesWatchedCallback

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 2 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > EnableCache --> CacheEnabled

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 3 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > repace 'enableCache' with Interop.WebContext.IsCacheEnabled(SwigCPtr);

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 4 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > removed.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 5 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > It should be a property, not a method.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 6 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > It should be a property, not a method.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 11 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > It should be a property, not a method.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 12 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > It should be a property, not a method.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 13 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs]: > It should be a property, not a method.

DONE

File: src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs (hunk 14 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs]: > ChangesWatchCallback --> ChangesWatchedCallback

DONE

File: src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs (hunk 15 above)
Comment (zhouleonlei) [reply src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs]: > RegisterChangesWatchCallback --> RegisterChangesWatchedCallback

DONE

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs (hunk 9 above)
Comment (huayongxu) [reply src/Tizen.NUI/src/internal/WebView/WebContext.cs:513]: please skip this.

**Review by AdunFang — APPROVED** [review]
It is OK.

**Review by elishateng — APPROVED** [review]
LGTM

File: src/Tizen.NUI/src/internal/WebView/WebContext.cs [hunk 16]
```diff
@@ -155,25 +191,151 @@ public string CertificateFilePath
...
+                Interop.WebContext.EnableCache(SwigCPtr, value);
+                if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
+            }
+        }
+
+        /// <summary>
+        /// Get context certificate file.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public string ContextCertificateFile
```
Comment (rabbitfor) [inline src/Tizen.NUI/src/internal/WebView/WebContext.cs:221]: Does it return a file name? (With full path?) Or does it return the content of certificate file in string?

File: src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs [hunk 17]
```diff
@@ -108,5 +117,17 @@ public void ClearCookies()
             Interop.WebCookieManager.ClearCookies(SwigCPtr);
             if (NDalicPINVOKE.SWIGPendingException.Pending) throw NDalicPINVOKE.SWIGPendingException.Retrieve();
         }
+
+        /// <summary>
+        /// Registers callback for changing watch.
+        /// <param name="callback">callback for changing watch</param>
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public void RegisterChangesWatchedCallback(ChangesWatchedCallback callback)
```
Comment (rabbitfor) [inline src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs:126]: What does it mean "Changes watched"?
"Changed detected" or "Watch changed"?

Comment (rabbitfor) [reply src/Tizen.NUI/src/internal/WebView/WebCookieManager.cs:126]: If it is similar with [this](https://developer.chrome.com/docs/extensions/reference/cookies/#event-onChanged), I think it should be an event and the `Changed` is better name.

**Comment by zhouleonlei** [comment]
Changed ContextCertificateFile  to CertificateFilePath
Changed  RegisterChangesWatchedCallback to RegisterWatchChangedCallback

**Comment by rabbitfor** [comment]
Hi Mr. Lei.
Regarding `Watch changed`, I still don't get it what it means.
Are there "Watch" somewhere? Or the cookie manager can watch something?

**Review by zhouleonlei — COMMENTED** [review]
> Hi Mr. Lei.
> Regarding `Watch changed`, I still don't get what it means.
> Is there a "Watch" somewhere? Or the cookie manager can watch something?

Hello Ms.Yang:
The EWK API name looks confused :). It means:
Watch for cookies' changes. Callback function that will be called every time cookies are added, removed or modified.
Should I change the name to "CookiesChangedCallback"?

**Comment by rabbitfor** [comment]
> > Hi Mr. Lei.
> > Regarding `Watch changed`, I still don't get what it means.
> > Is there a "Watch" somewhere? Or the cookie manager can watch something?
> 
> Hello Ms.Yang:
> The EWK API name looks confused :). It means:
> Watch for cookies' changes. Callback function that will be called every time cookies are added, removed or modified.
> Should I change the name to "CookiesChangedCallback"?

I think it'd better to be "CookieChanged" event. Not a callback. It is really kind of an event.

**Review by rabbitfor — APPROVED** [review]
LGTM

**Pull request merged**